### DISCUSSION TITLE
[Unstable] Creature ragdolls no longer refresh on spawning them

### DISCUSSION THREAD

**Discussion by ubertpendragon** [FakeFishGames/Barotrauma#13789] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [X] My issue happened while using mods.

### What happened?

In stable, you can simply spawn a new creature or select the creature again in creature editor to refresh its ragdoll. 
You do not need to go back to main menu and disable/enable mod. 

In Unstable this is no longer the case. You have to go back to the menu and manually refresh

Needless to say going to the menu slows down modding considerably. Reloadpackage is an option, but is still considerably slower than before and removes previous existing creatures. 

Unless there are very good reasons not to do so, packages should update as much as possible without reloading

### Version

v1.4.3.0 (unstable)

**Comment by NilanthAnimosus** [comment]
Thank you for the report!

Ragdolls and animation files are now cached ingame (I tested this myself actually), to my knowledge before when spawning characters there would be a slight freeze as it loads the character for each you spawn (mainly noticable on events that spawn several), which this adressed/improved by caching them on first spawn so there is reason to it.

For in game testing you can actually use reloadpackage mod force (Not the best idea for all cases, but creatures worked well enough). This should reload the package and yes, removes existing creatures. but before those existing creatures wouldn't have reloaded their ragdolls until you spawned new ones, which I can see some purpose too but the same can be achieved with two separate mods if your wanting to compare an old and new version (though I agree, thats slower and more work) but if you can suggest an alternative feel free (maybe a toggle in the games config like verbose mode to simply refresh every time instead of caching, even though its a performance detriment) - depends what @itchyOwl thinks.

Im not sure on the character editor personally (I am not well versed in it), but it does sound like perhaps something should change in the character editor (it may be possible to add a button or simply on selecting the character refresh / remove from both caches as if a mod reload occurs perhaps.

As a further note, it would be odd if we allowed one to edit their character ragdoll/animation files in a multiplayer session by it always reloading like the old behaviour as before that would likely have allowed the changes, but now since their cached it won't effect the game and on a rejoin to reload the changes fail the XML checks. so there is that too.

**Reply by ubertpendragon** [reply]
IF there's a performance gain, Reloadpackage it is; but it's definitely less convenient. 

> there would be a slight freeze as it loads the character for each you spawn (mainly noticable on events that spawn several)

On that note the Hammerhead Matriach still spawns 30 fishies on a single frame, causing a lag spike (despite the changes). Might want to use multiple delayed statuseffects

**Reply by itchyOwl** [reply]
The ragdolls were cached also previously, but caching just didn't work for variants. I'm not entirely sure what's the issue here, but there should be a shared ragdoll params (=prefab) per each species. And if you edit that ragdoll (params/prefab), it should affect all species using that. You shouldn't have to reload the package. So this could be a new issue. Need to look into it.

**Comment by Regalis11** [comment]
@ubertpendragon could you clarify at what point exactly are you expecting the ragdoll to get refreshed? Do you mean you are modifying the character in the character editor, and don't see the changes when it spawns in-game, or are you modifying the xml files manually?

**Reply by ubertpendragon** [reply]
If you edited the creatureragdoll.xml, freshly spawned creatures would inherit the ragdoll, but NOT the creature.xml
Previously, you'd would either:

* Type `spawn character cursor` while testing, freshly spawned creature with fresh ragdoll

* Use the drop down menu in character editor, pick a different character then back to your character to quickly refresh it (typically arrow down => arrow up). You can click refresh ragdoll but it's not as fast.
![image](https://github.com/FakeFishGames/Barotrauma/assets/73229309/b4c22504-3ebb-4ace-ad76-108d20a53d5d)

This was convenient as you could immediately change a bit of xml, spawn a creature and test/visualize the changes

**Reply by ubertpendragon** [reply]
Needless to say, being able to spawn "fresh" items/events without reloading the package every time via console would be a massive time saver.

**Reply by Regalis11** [reply]
So to be clear, you mean the changes don't get refreshed automatically if you edit the xml manually?

**Reply by ubertpendragon** [reply]
Newly spawned creatures used the latest version of the ragdoll.xml, without reloading the mod. 

It didnt update existing creatures

**Reply by Regalis11** [reply]
And you are editing the xml manually, not using the character editor?

**Reply by itchyOwl** [reply]
If the issue is that you aren't able to edit the xml "on the fly", it can be solved in the character editor by pressing the red "reset" button. In-game, you could use the "resetragdoll" console command.

**Reply by ubertpendragon** [reply]
Resetragdoll does not allow spawning creatures with fresh xml while testing

`reloadpackage [mod] force` can do it but it will delete all existing items/creatures previously placed

**Reply by ubertpendragon** [reply]
Long story short if it's for the greater good we can cope with reloadpackage, but otherwise the less commands/menus/restarts the better

**Reply by itchyOwl** [reply]
Could you clarify your workflow a bit? What do you mean by "spawning creatures with fresh xml while testing"? It seems that you are doing something I haven't even thought of 😅

**Reply by itchyOwl** [reply]
You might have benefitted from some bug that we've now fixed, but maybe there's a way to get back the old behavior too. Just would have to understand first what it is.

**Reply by ubertpendragon** [reply]
> Could you clarify your workflow a bit? What do you mean by "spawning creatures with fresh xml while testing"? It seems that you are doing something I haven't even thought of

1. Open sub editor
2. Test
3. `spawn creature cursor`
4. edit creatureragdoll.xml and save (change limb colors for testing)
5. `spawn creature cursor`
6. enjoy the freshly updated creature ragdoll without going through main menu or console

**Reply by NilanthAnimosus** [reply]
You could use the bindkey to make it a single keypress to reload your mod with by having that execute the command. if that helps to speed up the process a bit without having to open the console.

I assume what you want to do is have the old ragdoll and new ragdoll simultaniously present (by spawning new characters with modified ragdolls each time) and possibly comparing the result as you modify (?). at least for understanding what your after. with minimal amount of game freezes to each individual tweak (such as reloadpackage).

**Reply by ubertpendragon** [reply]
I'll probably just close this now unless somebody else complains about it. 

Everyone has a different workflow I guess, I just figured this was the fastest in terms of modding creatures

**Reply by itchyOwl** [reply]
Both creatures in that example should use the same ragdoll params. "resetragdoll" should load the changes from xml, but it seems you'll have to run the command for each character individually (should change this). You said that it doesn't "allow spawning characters with fresh ragdoll". Do you mean that you'd want to spawn characters without the changes you made in the xml? I just tested and this didn't happen earlier on regular characters, but I think it could have happened on variants, because variant caching was broken.

If you want to experiment with some ragdoll changes that aren't saved, I's suggest using the character editor, because that's what's it for. Can't handle everything, but most things it can.

**Reply by itchyOwl** [reply]
Fixed the "resetragdoll" command in FakeFishGames/Barotrauma-development#5224. It now lists species names and updates all the character instances of that species.
